File: MAQ Software/467.unique-substrings-in-wraparound-string.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

class Solution {
public:
    int findSubstringInWraproundString(string s) {
        vector<int> best(26, 0);
        int run = 0;
        char prev = 0;
        for (char c : s) {
            if (prev && (c - prev + 26) % 26 == 1) ++run;
            else run = 1;
            best[c - 'a'] = max(best[c - 'a'], run);
            prev = c;
        }
        return accumulate(best.begin(), best.end(), 0);
    }
};

/*
Interview explanation:
For each ending letter, only the longest valid wraparound substring ending there matters. It contributes all shorter suffix lengths ending at that letter without duplicates.

C++ data structures: a fixed vector<int> of size 26 stores maximum run lengths.

Edge cases: z->a is valid through modulo arithmetic; breaks reset the run to 1.

Complexity: O(n) time and O(1) space.
*/

File: companies/MAQ Software/467.unique-substrings-in-wraparound-string.java
/**
 * Algorithm:
 * Track the current consecutive run in the infinite wraparound alphabet. For
 * each ending character, only the maximum run length matters because it
 * represents all unique substrings ending at that character.
 *
 * Java data structures:
 * A fixed int[26] stores the best run length by ending character.
 *
 * Complexity:
 * Time O(n), space O(1).
 */
class Solution {
    public int findSubstringInWraproundString(String s) {
        int[] best = new int[26];
        int run = 0;
        char prev = 0;
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (prev != 0 && (ch - prev + 26) % 26 == 1) {
                run++;
            } else {
                run = 1;
            }
            int idx = ch - 'a';
            best[idx] = Math.max(best[idx], run);
            prev = ch;
        }
        int ans = 0;
        for (int x : best) {
            ans += x;
        }
        return ans;
    }
}

